МИРЭК | 4 модуль

*Автор: Татьяна Рогович*

# Основы программирования в Python
## Cеминар 7

## Введение в ML. Задача регрессии. Решающее дерево и случайный лес

Этот блокнот составлен на основе большого блокнота, который доступен по этой [ссылке](https://www.kaggle.com/dansbecker/your-first-machine-learning-model). В нем присутствуют дополнительные комментарии и другие примеры работы алгоритма. Если хотите как следует разобраться с деревьями на базовом уровне - обязательно читайте.

### Загружаем наши данные

Нашим датасетом будет информация о домах, которая включает в себя множество параметров, например, количество и размер этажей дома, год постройки, общая площадь, цена и многое другое (остальные столбцы, можете изучить самостоятельно).

In [2]:
import pandas as pd

#Загружаем данные
home_data = pd.read_csv('https://raw.githubusercontent.com/rogovich/2019-2020_PolSci_Data_Analysis_in_Python/master/12week_ML_Intro/house_data.csv')

### Выбор целевой переменной

В нашем случае, все более-менее очевидно. Это цена на дом. Обычно, цель предсказания помечают y.

In [3]:
y_train = home_data.SalePrice

### Выбор признаков (Features)

Столбцы, которые есть в нашей модели и которые в последствии будут использованы для предсказания, называются признаками (features). В нашем случае, эти колонки будут определять стоимость дома. Иногда используются все колонки, кроме той на которую делается предсказание. А вдругих случаях, лучше выбрать только часть из них. Как определиться? Для этого проводят разведывательный анализ или сравнивают качество работы моделей на разных выборках признаков.

Мы будет использовать не все столбцы, чтобы алгоритм все считал быстрее. Добавим в модель только те столбцы, которые сильнее всего коррелировали с ценой, когда мы исследовали данные.

In [4]:
feature_columns = ['OverallQual', 'GrLivArea', 'GarageArea', 'TotalBsmtSF', 'FullBath', 'YearBuilt', 'YearRemodAdd']

Обычно, такие данные обозначаются X:

In [22]:
X_train = home_data[feature_columns]

### Строим модель

Мы будем использовать модуль scikit-learn для создания модели. Scikit-learn одна из самых популярных библиотек для моделирования данных, хранящихся в датафреймах. 

Для построения модели нужно выполнить следующие шаги: <br>

Define: Какого типа будет модель? Дерево решений? Какой-то другой тип?<br>
Fit: запускаем модель на тренировочных данных (обучаем ее), чтобы алгоритм нашел в них некие закономерности и зависимости.<br>
Predict: Предсказать результат.<br>
Evaluate: Определить насколько точным оказалось предсказывание, оценить качество модели<br>

В нашем примере, мы будет использовать дерево решений из scikit-learn и тренировать модель с признаками, выделенными ранее. Импортируем функцию DecisionTreeRegressort

In [23]:
from sklearn.tree import DecisionTreeRegressor

Инициализируем модель и обучим ее на наших данных.

In [24]:
# Инициалазируем модель
home_model = DecisionTreeRegressor()

# Обучаем модель
home_model.fit(X_train, y_train)

DecisionTreeRegressor()

Нам выше вывелась информация о нашем регрессоре и с какими парамтерами он обучался.

Мы обучили модель. Но как хорошо она обучилась? Давайте предскажем значения и сравним их с теми ценами, которые на самом деле соответствуют домам.

In [25]:
print("Наше предсказание:", home_model.predict(X_train.tail(10)))
print("Реальная цена домов:", y_train.tail(10).tolist())

Наше предсказание: [136000. 287090. 145000.  84500. 185000. 175000. 210000. 266500. 142125.
 147500.]
Реальная цена домов: [136000, 287090, 145000, 84500, 185000, 175000, 210000, 266500, 142125, 147500]


Результаты похожи, но кое-где есть отличия. И мы отображаем всего 10 элементов из всей выборки. Посмотрим насколько правильно нам предсказала модель для всей выборки. Мы можем, конечно, пройтись в цикле по всем эелементам и посчитать ошибку у каждого. Но у нас могут быть как хорошие, так и плохие предсказания. И в идеальном случае нам нужна одна метрика, которая скажет как хорошо мы обучили данные.

Воспульзуемся метрикой MEA (Mean Absolute Error) - Cредняя Aбсолютная Ошибка. Ее можно представить в таком виде: ошибка = реальная цена − предсказанная цена. Например, если цена дома 150 000, мы предсказали цены в 100 000, то ошибка будет 50 000. Для ошибок в отрицательную сторону берем модуль. Это одна из простейших метрик.

Импортируем нужную функцию опять же из sklearn

In [9]:
from sklearn.metrics import mean_absolute_error

Делаем предсказание и передаем в функцию предсказанное и реальное значение

In [10]:
pred = home_model.predict(X_train)
mae = mean_absolute_error(pred, y_train)
print(mae)

111.4013698630137


Получили ошибку достаточно маленькую ошибку цены в долларах, что было бы очень хорошим результатом, если не одно но.

### Валидация данных

Мы только что проверяли обученную модель на тех же данных, на которых мы ее и обучали. И это неверно. Т.к. в нашем случае нам нужно предсказывать цены новых домов, которых не было в обучающей выборке. Поэтому не удивительно, что MAE = 111. 

Опишем эту проблему таким примером. У нас есть большой рынок жилья. Цена жилья не связана с цветом входной двери. Но в нашей обучающей выборке у всех домов с высокой ценой была зеленая дверь. Модель увидит эту зависимость и будет предсказаывать для новых домов с зеленой дверью высокую цену, что не будет корректным. Но при валидации модели на обучающих данных мы получили маленькую ошибку и посчитали, что модель натренирована хорошо. Если бы мы это сделали на тестовых данных, то увидели бы что ошибка велика и нужно переделать модель.

### Обучающие и тестовые выборки

Давайте разделим наши данные на 2 части: обучающую выборку и тестовую. Воспльзуемся опять модулем sklearn и функцией train_test_split.

In [26]:
from sklearn.model_selection import train_test_split

Разобъем данные как для X, так и для  y. Каждый раз разбитие происходит случайно, мы можем только указать в какой пропорции мы хотим разбить наши данные, например 30% всей выборки на обучение, остальное на тест.

In [29]:
# заполнили массивы с нашими данными 
X = home_data[feature_columns]
y = home_data.SalePrice

# Делим
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size = 0.25)

Посмотрим сколько данных попала в обучающую и тестовые выборки

In [30]:
print(len(train_X))

1095


In [31]:
print(len(test_X))

365


В обучающей 1095, а в тестовой 365 домов.

Теперь как и в прошлый раз построим модель и посчитаем MAE

In [35]:
# Инициализируем модель
home_model = DecisionTreeRegressor()
# Обучаем модель
home_model.fit(train_X, train_y)

# Предсказываем цены и считаем MAE
val_predictions = home_model.predict(test_X)
print(mean_absolute_error(test_y, val_predictions))


26924.701369863014


Воу. Это большая ошибка по сравнению с нашей изначальной ошибкой на обучающей выборке. И выходит, что для новых домов, наша модель цены предсказывала бы плохо.

Наша модель обучилась не очень оптимально, но к счастью есть несколько способов как можно улучшить модель, например, экспериментируя с параметрами модели.

### Эксперементируем с моделями

Можно сказать что мы столкнулись с проблемой переобучения (overfitting), когда модель предсказывает обучающие данные практически идеально, а тестовые данные и другие данные предсказывате плохо.

Существует еще и недообучение (underfitting), когда модель не может найти параметры по которым можно хорошо разделить данные, и она плохо предсказывает даже на обучающих данных.

Так как мы используем DecisionTreeRegressor, т.е. дерево, в случае overfitting дерево обычно очень глубокое, а при underfitting получаем деревья небольшие. Поэтому мы может попробовать регулировать глубину дерева. В DecisionTreeRegressor есть параметер max_leaf_nodes, который мы можем контролировать. Чем больше листьев мы разрешим модели делать, тем дальше мы уйдем от underfiting, но будем приближаться к overfitting, т.е. в идеальном случае нам нужно подобрать примерно среднюю глубину дерева для модели.

Создадим функцию, которая будет строить для нас деревеья с различным параметром max_leaf_nodes. А замерять модели будем снова через MAE.

In [37]:
def get_mae(max_leaf_nodes, train_X, test_X, train_y, test_y):
    # инициализируем модель с параметром max_leaf_node, который мы передали в функцию
    model = DecisionTreeRegressor(max_leaf_nodes=max_leaf_nodes)
    # обучаем модель на обучающих данных
    model.fit(train_X, train_y)
    # предсказываем на тестовых данных
    preds_val = model.predict(test_X)
    # считаем MAE
    mae = mean_absolute_error(test_y, preds_val)
    # возвращем значение ошибки
    return(mae)

В функцию мы соответственно передаем количество листов, которые мы хотим и обучающие и тестовые выборки. Сделаем небольшой цикл по различный значением глубины от 5 до 5000 и выведем результаты.

In [41]:
# сравниваем различне значения MAE для различной глубины деревьев
for max_leaf_nodes in [5, 10, 50, 100, 500, 1000, 5000]:
    my_mae = get_mae(max_leaf_nodes, train_X, test_X, train_y, test_y)
    print("Максимальная глубина дерева: %d  \t\t MEA:  %d" %(max_leaf_nodes, my_mae))

Максимальная глубина дерева: 5  		 MEA:  35992
Максимальная глубина дерева: 10  		 MEA:  30972
Максимальная глубина дерева: 50  		 MEA:  26146
Максимальная глубина дерева: 100  		 MEA:  26710
Максимальная глубина дерева: 500  		 MEA:  27211
Максимальная глубина дерева: 1000  		 MEA:  26554
Максимальная глубина дерева: 5000  		 MEA:  26969


Как мы видим, самый лучший результат мы получили для глубины 100, но при этом ошибка достаточно большая и искажает предсказания для новых данных. С другой стороны, если мы посмотрим описательные статистики для нашей цены, мы увидим, что по сравнению со медианой и интерквартильным размахом, наша ошибка не настолько ужасна. Поэтому величина ошибки предсказания достаточно хитрая штука для интерпретации.

In [19]:
train_y.describe()

count      1095.000000
mean     181712.286758
std       77955.082565
min       34900.000000
25%      130000.000000
50%      165000.000000
75%      215000.000000
max      745000.000000
Name: SalePrice, dtype: float64

### Random Forest

Дерево решений оставляет с серьезным выбором. Глубокое дерево с overfitting с несколькими домами в каждом листе или мелкое дерево с несколькими листами, которое не может определить параметры и разделить данные? И все это с приличной ошибкой.

Некоторые другие модели деревьев смогли решить эту проблему и уменьшить при этом итоговую ошибку. Попробуем взять Random Forest. Он использует множество деревьев и делает предсказание путем усреднения прогнозов каждого дерева.

[Подробнее про случайный лес тут.](https://dyakonov.org/2016/11/14/%D1%81%D0%BB%D1%83%D1%87%D0%B0%D0%B9%D0%BD%D1%8B%D0%B9-%D0%BB%D0%B5%D1%81-random-forest/)

Попробуем использовать Random Forest на наших данных. Импортируем его из модуля sklearn

In [43]:
from sklearn.ensemble import RandomForestRegressor

Строим теперь модель с тем же параметрами и данными как в DecissionTreeRegressor

In [27]:
# Инициализируем модель
forest_model = RandomForestRegressor()
# Обучаем модель
forest_model.fit(train_X, train_y)
# Делаем предсказание
randomf_preds = forest_model.predict(test_X)
# Посчитаем и выведем MAE
print(mean_absolute_error(test_y, randomf_preds))

19441.526748858447


Ошибка уменьшилась, но все равно составляет больше 20 тысяч долларов. Можно продолжить дальше эксперементировать с моделью, изменяя параметры или увеличивая обучающую выборку. 

Давайте изменим параметр n_estimators, который отвечает за количество деревьев в модели ( не путать с количеством листьев в DecisionTree)

In [48]:
# Инициализируем модель
forest_model = RandomForestRegressor(n_estimators = 100)
# Обучаем модель
forest_model.fit(train_X, train_y)
# Делаем предсказание
randomf_preds = forest_model.predict(test_X)
# Посчитаем и выведем MAE
print(mean_absolute_error(test_y, randomf_preds))

21662.834916046966


Уже 19 тысяч! Много это или мало? Можно ли остановиться? 

Конечно, мы всегда стараемся минимизировать нашу ошибку, но это и не всегда возможно. Возможно, здесь лучше справится другая модель. Так же мы взяли очень мало признаков, при чем не основываясь на какой-то особой теории - возможно, преобразование наших признаков или же их замена, улучшили бы качество.

С остальными параметрами можно ознакомиться в [документации](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) функции

Дополнительным приятным бонусом использования моделей дерева и случайного леса является то, что мы можем использовать их определения важности признаков. Внутри себя дерево ранжируюет признаки на основании того, насколько разбиение по ним снижает неоднородность данных. Так мы можем использовать лес для отбора наиболее важных признаков.

In [29]:
list(zip(feature_columns, forest_model.feature_importances_))

[('OverallQual', 0.5767274638092772),
 ('GrLivArea', 0.1800254228003314),
 ('GarageArea', 0.057984302587197625),
 ('TotalBsmtSF', 0.09873909813622055),
 ('FullBath', 0.016522559961700576),
 ('YearBuilt', 0.03921901573357202),
 ('YearRemodAdd', 0.03078213697170065)]

In [34]:
print("Features sorted by their score:")
print(sorted(zip(map(lambda x: round(x, 4), forest_model.feature_importances_), feature_columns), 
             reverse=True))

Features sorted by their score:
[(0.5767, 'OverallQual'), (0.18, 'GrLivArea'), (0.0987, 'TotalBsmtSF'), (0.058, 'GarageArea'), (0.0392, 'YearBuilt'), (0.0308, 'YearRemodAdd'), (0.0165, 'FullBath')]


Давайте сравним с матрицей корреляции. Если бы мы ориентировались только на корреляции, мы считали два первых признака почти одинаково важным. По важностям, определенными решающим деревом видно, что 'OverallQual' с большим отрывом опережает площадь.

In [33]:
home_data[feature_columns + ['SalePrice']].corr()['SalePrice']

OverallQual     0.790982
GrLivArea       0.708624
GarageArea      0.623431
TotalBsmtSF     0.613581
FullBath        0.560664
YearBuilt       0.522897
YearRemodAdd    0.507101
SalePrice       1.000000
Name: SalePrice, dtype: float64

На следующем семинаре мы вернемся к "Титанику" и попробуем, наконец, предсказать, кто же выжил во время крушения.